Kernel/ACPI/MultiProcessorParser.h

```diff
@@ -0,0 +1,241 @@
+    };
+
+    enum class InterruptType {
+        INT = 0,
+        NMI = 1,
+        SMI = 2,
+        ExtINT = 3,
+    };
+
+    struct [[gnu::packed]] IOInterruptAssignmentEntry
+    {
+        EntryHeader h;
+        u8 interrupt_type;
+        u8 polarity;
+        u8 trigger_mode;
+        u8 source_bus_id;
+        u8 source_bus_irq;
+        u8 destination_ioapic_id;
+        u8 destination_ioapic_intin_pin;
+    };
+
+    struct [[gnu::packed]] LocalInterruptAssignmentEntry
+    {
+        EntryHeader h;
+        u8 interrupt_type;
+        u8 polarity;
+        u8 trigger_mode;
+        u8 source_bus_id;
+        u8 source_bus_irq;
+        u8 destination_lapic_id;
+        u8 destination_lapic_lintin_pin;
+    };
+
+    enum class SystemAddressType {
+        IO = 0,
+        Memory = 1,
+        Prefetch = 2,
+    };
+
+    struct [[gnu::packed]] SystemAddressSpaceMappingEntry
+    {
+        ExtEntryHeader h;
+        u8 bus_id;
+        u8 address_type;
+        u64 address_base;
+        u64 length;
+    };
+
+    struct [[gnu::packed]] BusHierarchyDescriptorEntry
+    {
+        ExtEntryHeader h;
+        u8 bus_id;
+        u8 bus_info;
+        u8 parent_bus;
+        u8 reserved[3];
+    };
+
+    struct [[gnu::packed]] CompatibilityBusAddressSpaceModifierEntry
+    {
+        ExtEntryHeader h;
+        u8 bus_id;
+        u8 address_modifier;
+        u32 predefined_range_list;
+    };
+
+}
+
+class PCIInterruptOverrideMetadata : public RefCounted<PCIInterruptOverrideMetadata> {
+public:
+    PCIInterruptOverrideMetadata(u8 bus_id, u8 polarity, u8 trigger_mode, u8 source_irq, u32 ioapic_id, u16 ioapic_int_pin);
+    u8 bus();
+    u8 polarity();
+    u8 trigger_mode();
+    u8 pci_interrupt_pin();
+    u8 pci_device_number();
+    u32 ioapic_id();
+    u16 ioapic_interrupt_pin();
+
+private:
+    u8 m_bus_id;
+    u8 m_polarity;
+    u8 m_trigger_mode;
+    u8 m_pci_interrupt_pin;
+    u8 m_pci_device_number;
+    u32 m_ioapic_id;
+    u16 m_ioapic_interrupt_pin;
+};
+
+class MultiProcessorParser {
+public:
+    static MultiProcessorParser& the();
+
+    static bool is_initialized();
+    static void initialize();
+    Vector<RefPtr<PCIInterruptOverrideMetadata>>& get_pci_interrupt_redirections();
+
+protected:
+    MultiProcessorParser();
+
+    void parse_configuration_table();
+    size_t get_configuration_table_length();
+    void parse_floating_pointer_data();
+
+    Vector<unsigned>& get_pci_bus_ids();
+
+    MultiProcessor::FloatingPointer* search_floating_pointer();
+    MultiProcessor::FloatingPointer* search_floating_pointer_in_ebda(u16 ebda_segment);
+    MultiProcessor::FloatingPointer* search_floating_pointer_in_bios_area();
+
+    MultiProcessor::FloatingPointer* m_floating_pointer;
+    MultiProcessor::ConfigurationTableHeader* m_configuration_table;
+    Vector<MultiProcessor::IOInterruptAssignmentEntry*> m_io_interrupt_redirection_entries;
+    Vector<MultiProcessor::BusEntry*> m_bus_entries;
```

Comment:
If I understand correctly, these are all offsets, not actually pointers.
As such, we should not store them as pointers, as it's not safe to dereference them.